### PR TITLE
Check cause of exception for retry in Elastic sink [HZ-2335]

### PR DESCRIPTION
**Pull request by frant-hartm**
Elasticsearch started wrapping some IOExceptions into ElasticsearchException somewhere between 7.10.x and 7.17.x.

Added a test similar to retry test in IT test suite, but for sink.

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases

Should be backported to 5.0.z, 5.1.z, 5.2.z

### CODE REVIEW

File: extensions/elasticsearch/elasticsearch-7/src/test/java/com/hazelcast/jet/elastic/RetryElasticSinkTest.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import org.junit.Test;
+import org.testcontainers.containers.ToxiproxyContainer;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.elasticsearch.client.RequestOptions.DEFAULT;
+
+/**
+ * Test running single Jet member locally and Elastic in docker
+ */
+public class RetryElasticSinkTest extends BaseElasticTest {
```
Comment (olukas) [inline extensions/elasticsearch/elasticsearch-7/src/test/java/com/hazelcast/jet/elastic/RetryElasticSinkTest.java:48]: Could you move it to nightly tests?

**Review by olukas — CHANGES_REQUESTED** [review]

File: extensions/elasticsearch/elasticsearch-7/src/main/java/com/hazelcast/jet/elastic/impl/RetryUtils.java [hunk 2]
```diff
@@ -77,6 +77,10 @@ private static boolean anyOf(Exception e, Class<? extends Exception>[] exception
             if (exception.isAssignableFrom(e.getClass())) {
                 return true;
             }
+            if (e.getCause() != null && e.getCause() instanceof Exception && e.getCause() != e) {
+                Exception cause = (Exception) e.getCause();
+                return anyOf(cause, exceptions);
+            }
```
Comment (TomaszGaweda) [inline extensions/elasticsearch/elasticsearch-7/src/main/java/com/hazelcast/jet/elastic/impl/RetryUtils.java:83]: You can extend ExceptionUtil.isOrHasCause

File: extensions/elasticsearch/elasticsearch-7/src/test/java/com/hazelcast/jet/elastic/RetryElasticSinkTest.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    @Test
+    public void when_elasticNotInitiallyAvailable_then_shouldWriteAllDocuments() throws Exception {
+        int batchSize = 10_000;
+        TestItem[] items = new TestItem[batchSize];
+        for (int i = 0; i < batchSize; i++) {
+            items[i] = new TestItem("id" + i, "name" + i);
+        }
+
+        ToxiproxyContainer.ContainerProxy elasticProxy = toxiproxy.getProxy(ElasticSupport.elastic.get(), 9200);
```
Comment (TomaszGaweda) [inline extensions/elasticsearch/elasticsearch-7/src/test/java/com/hazelcast/jet/elastic/RetryElasticSinkTest.java:76]: ContainerProxy is deprecated, please use new API, something like:
```java
        final ToxiproxyClient toxiproxyClient = new ToxiproxyClient(toxi.getHost(), toxi.getControlPort());
        final Proxy proxy = toxiproxyClient.createProxy("elastic", "0.0.0.0:8670", "elastic:9200");
```

and then:
```
proxy.toxics().timeout("TIMEOUT", UPSTREAM, 0);
// ...
proxy.toxics().get("TIMEOUT").remove();
```

**Review by TomaszGaweda — CHANGES_REQUESTED** [review]

File: extensions/elasticsearch/elasticsearch-7/src/test/java/com/hazelcast/jet/elastic/RetryElasticSinkTest.java (hunk 1 above)
Comment (frant-hartm) [reply extensions/elasticsearch/elasticsearch-7/src/test/java/com/hazelcast/jet/elastic/RetryElasticSinkTest.java:55]: 31b46fb8fc1

File: extensions/elasticsearch/elasticsearch-7/src/test/java/com/hazelcast/jet/elastic/RetryElasticSinkTest.java (hunk 3 above)
Comment (frant-hartm) [reply extensions/elasticsearch/elasticsearch-7/src/test/java/com/hazelcast/jet/elastic/RetryElasticSinkTest.java]: 31b46fb8fc1

**Comment by frant-hartm** [comment]
Unrelated test failure.

**Pull request merged**